Get response set from dota2 wiki

choldener opened the pull request:
Noticed on config.py there was a todo getting the hero and item response set from the wiki, quickly created a solution that can keep the bot more up-to-date.

First pull request so apologies if formatting is not standard.

choldener committed:
feat: get response set from wiki

Using urllib.request and json package, automated the item and hero sets.

choldener committed:
Fix: Turned keys lowercase

Keys should be lowercase

choldener committed:
Fix: Remove special characters from items

Keys should be without special characters.

choldener commented:
Not sure how you would want to handle items from previous patches such as 'blades of attack  pre 6.54 '

choldener committed:
Fix: Removed special characters from hero names

included special characters in for statement.

choldener committed:
Feat: Added exclusion set for heroes/items

MePsyDuck commented:
Hey, thanks for taking time to do this. The changes seems fine, but I feel like the core logic can be moved into separate file like util so as to keep config free of any complex logic. This way, we can also move the logic into a method and call it for both items and heroes.
I would also suggest you to use `preprocess_text` in `str_utils` to remove special characters, and use `requests` in place of `urllib`.

Jonarzz commented on config.py:
```diff
@@ -80,56 +82,32 @@
+    HERO_NAME_RESPONSES = set()
+    for i in json.loads(url.read().decode())['cargoquery']:
+        i = i['title']['title'].lower()
+        for c in SPECIAL_CHARACTERS: i = i.replace(c, " ")
```
Please, replace the one line `for`s and `if`s with ones with line breaks and formatting to increase readability.

choldener replied:
fixed in https://github.com/Jonarzz/DotaResponsesRedditBot/pull/41/commits/8b9847827e79909c6543fbb1c3b459ed68bf6d48

Jonarzz commented on config.py:
```diff
@@ -80,56 +82,32 @@
+                            '+IS+NULL&limit=500&format=json') as url:
+    SELECTED_HERO_RESPONSES = {}
+    HERO_NAME_RESPONSES = set()
+    for i in json.loads(url.read().decode())['cargoquery']:
```
`i` does not say much about what the variable is, would be better to name it in a more verbose way for better readability and future understanding of the code.

choldener replied:
I have changed around the variable names to hopefully be more verbose and to increase readability.  Please let me know if there is any confusion with the names and I can find something that works better. https://github.com/Jonarzz/DotaResponsesRedditBot/pull/41/commits/8b9847827e79909c6543fbb1c3b459ed68bf6d48

Jonarzz commented on config.py:
```diff
@@ -80,56 +82,32 @@
+                            'title&where=game+IS+NULL&limit=500&format=json') as url:
+    SELECTED_ITEM_RESPONSES = {}
+    ITEM_RESPONSES = set()
+    for i in json.loads(url.read().decode())['cargoquery']:
```
Same as above.

Jonarzz reviewed (changes requested):
Please, apply the changes requested by @MePsyDuck above first.

choldener committed:
refactor: Moved logic from config to response_request.py

Improved variable names
Included line breaks in for and if statments
replaced urllib/json with requests

MePsyDuck commented on util/response_request.py:
```diff
@@ -0,0 +1,30 @@
+                                'action=cargoquery&tables=heroes&fields=title&where=game'+
+                                '+IS+NULL&limit=500&format=json')
+    WEB_JSON = WEB_REQUEST.json()
+    SELECTED_HERO_RESPONSES = {}
```
What purpose does SELECTED_HERO_RESPONSES serve here?

choldener replied:
In the comments I saw that responses were hard coded to increase flexibility. This is just an additional set that can be added to in order to exclude heroes or items (heros in this case) from being included in the set. 

Say you wanted a hero name to not be in your excluded set, you would add it there. Not needed but thought it could help address any flexibility issues that could arise in the future.

choldener replied:
Removed in https://github.com/Jonarzz/DotaResponsesRedditBot/pull/41/commits/60c457340a9fc231d6bd8ad0387d2898d82ce3a1

When ahead and removed the feature. It was a novel idea but I think it adds unneeded complexity and if an issue ever arises it can be reincorporated.

MePsyDuck commented on config.py:
```diff
@@ -80,56 +81,10 @@
-                       'sniper', 'oracle', 'timbersaw', 'bristleback', 'snapfire', 'void spirit'}
+# Hero and item responses not hardcoded here
+# Drawbacks: Can't be tweaked
+HERO_NAME_RESPONSES = request_hero_name_set()
```
`request_cargo_set` can be one method with the url as parameter, as rest of the logic is the same.

choldener replied:
Good catch, fixed in https://github.com/Jonarzz/DotaResponsesRedditBot/pull/41/commits/c370f726561bfe9b60a2ec63a0d868aa89905b8a.

MePsyDuck commented on util/response_request.py:
```diff
@@ -0,0 +1,30 @@
+
+
+def request_hero_name_set():
+    WEB_REQUEST = requests.get('https://dota2.gamepedia.com/api.php?' +
```
As a general rule of thumb, variables names shouldn't be uppercase unless they are constant and final (or atleast what I follow in Java for static final variables). Variables used only for inner method logic should be in snake_case.

choldener replied:
Fixed in https://github.com/Jonarzz/DotaResponsesRedditBot/pull/41/commits/c370f726561bfe9b60a2ec63a0d868aa89905b8a.

MePsyDuck commented on util/response_request.py:
```diff
@@ -0,0 +1,30 @@
+        HERO_ENTRIES = preprocess_text(WEB_OBJECT['title']['title'])
+        if HERO_ENTRIES not in SELECTED_HERO_RESPONSES: 
+            HERO_NAME_RESPONSES.add(HERO_ENTRIES)
+    return (HERO_NAME_RESPONSES)
```
` return (HERO_NAME_RESPONSES) ` is same as ` return HERO_NAME_RESPONSES `, so the extra paranthesis can be dropped.

choldener replied:
Fixed in https://github.com/Jonarzz/DotaResponsesRedditBot/pull/41/commits/c370f726561bfe9b60a2ec63a0d868aa89905b8a.

MePsyDuck commented:
On another note, this PR is for my dev branch, whereas is should've been for master.

choldener commented:
> On another note, this PR is for my dev branch, whereas is should've been for master.

Quick git question, the reason I chose your dev branch was because you are maintaining the code and made changes to the config file on this branch that was not represented on the master branch. Is there no conflict when that happens or what should be the main approach in situations like these.  I'm pretty new to collaborate github projects so pardon my ignorance.

Also should I go ahead and change from your branch to the master branch?

choldener committed:
Refactor: Condensed function and fixed variables

snake_case on variables
condensed both requests to a single request_cargo_set function

choldener committed:
Removed excluded set in cargo request

 Removed Feature: Added unneeded complexity and decreased readbility.

Jonarzz commented:
I think that the changes should be merged to the master to make all PRs more obvious and baseline clear. If there is a conflict, it should be resolved in the PR that comes/is to be accepted later. Any other dev branches should be kept up to date and rebased on master before and during development.
